### ISSUE TITLE
Commit message in OCA repositories

### ISSUE THREAD

**Issue by pedrobaeza** [OCA/oca-addons-repo-template#172]
@simahawk and me talked at Odoo XP about improving the commit messages on OCA repos, whose quality, being nice, is not the best one.

There are 2 open fronts here:

- Determine a good commit message structure. Simone is not very comfortable spending 5 chars to define the type tag (`[FIX]`, `[MIG]`...) in order to comply with the recommended first line length. One possibility is to short them (`[F]` or `FIX` without the brackets), but we still have some spending. The other is to put such tag on the expanded explanation on the second line.
- Enforce this structure through a linter. For this, one good option is https://commitizen-tools.github.io/commitizen/, that can be plugged into pre-commit.

What are your thoughts on this?

**Comment by legalsylvain** [comment]
"[F]" vs "[FIX]" vs "FIX" : hum... it's a matter of taste for me.

I am not in favour of creating yet another OCA rule (especially lint checks) which could prevent contributions or make merge slower.

**Comment by sbidoul** [comment]
I'd rather not invent a new convention. Either we stick to the [Odoo convention](https://www.odoo.com/documentation/15.0/contributing/development/coding_guidelines.html#tag-and-module-name), or with use something else that exists such as [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).

**Comment by pedrobaeza** [comment]
Yeah, Sylvain, I know each rule is a potential barrier, but not enforcing this provokes a nightmare to do a git log/git blame to know the content (or even the affected module) of the commit. Commit message rules can be very soft (only conforms Tag + module + description).

**Comment by pedrobaeza** [comment]
And about the second point, Stéphane?

**Comment by sbidoul** [comment]
Not sure. I've never tried enforcing commit message structure. That said I also think commit messages should at least include de module name. PR titles too :) But some commits don't qualify (e.g. dotfiles updates, etc). And what about fixups?
So yeah, does not sound easy to do...

**Comment by MiquelRForgeFlow** [comment]
Moreover, about the linter. If you do a module migration with 200 commit history, and many of those commits are linter-wrong, what do you do? How to let know to the pre-commit which commit has to pass the linter and which don't?

**Comment by sbidoul** [comment]
/me adding this to the list of arguments to change the migration process :grin:

**Comment by pedrobaeza** [comment]
> /me adding this to the list of arguments to change the migration process grin

The problem will exist the same if there are pending commits since the branch creation. But the idea here is to enforce only the strictly needed. Maybe through a commit date filter or similar.

**Comment by etobella** [comment]
One option is to add a configuration file in repos after creating a repo example and let people start using it. It is less agressive and this way we can check how to do it. For the configuration, we need something like https://github.com/apheris/cz-github-jira-conventional where all the logic can be defined there

**Comment by etobella** [comment]
I made a small demo test, really interesting :thinking: 
https://github.com/tegin/cz-oca

**Comment by grindtildeath** [comment]
I just gave another look at: [7.1   Commit message](https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#71commit-message)

There are good things in there, and I think we should encourage people to improve their messages but I agree with @legalsylvain that it could prevent people from contributing, so I'm against too much enforcing.

Personally I don't like having too many prefixes, and IMO if we have the type of change AND the module name (which can sometimes be quite long) we won't have enough room for a proper commit message.
I'd say if we have to choose between keeping the type of change or the module name in the first line, I'd prefer having to use the type there and write the module name in the message's "body" on third line, although IMO it shouldn't be needed as we could always check which module was updated according to the changed files, and it also allows more characters to write a proper message on the first line.

Finally I would prefer not being too strict and accept different "styles" as long as the most important information (i.e. what it's doing) is written down.
